lightning/src/offers/offer.rs

```diff
@@ -0,0 +1,778 @@
+use util::ser::{Writeable, Writer};
+
+use prelude::*;
+
+#[cfg(feature = "std")]
+use std::time::SystemTime;
+
+/// Builds an [`Offer`] for the user-pays-merchant flow.
+///
+/// See [module-level documentation] for usage.
+///
+/// [module-level documentation]: self
+pub struct OfferBuilder {
+	offer: OfferContents,
+}
+
+impl OfferBuilder {
+	/// Creates a new builder for an offer with the given description and node id.
+	pub fn new(description: String, node_id: PublicKey) -> Self {
+		let offer = OfferContents {
+			chains: None, metadata: None, amount: None, description, features: None,
+			absolute_expiry: None, issuer: None, paths: None, quantity_min: None,
+			quantity_max: None, node_id: Some(node_id),
+		};
+		OfferBuilder { offer }
+	}
+
+	/// Sets a chain hash of the given [`Network`] for the offer. If not called,
+	/// [`Network::Bitcoin`] is assumed.
+	///
+	/// Successive calls to this method will add another chain hash.
+	pub fn chain(mut self, network: Network) -> Self {
+		let chains = self.offer.chains.get_or_insert_with(Vec::new);
+		let block_hash = ChainHash::using_genesis_block(network);
+		if !chains.contains(&block_hash) {
+			chains.push(block_hash);
+		}
+
+		self
+	}
+
+	/// Sets the metadata for the offer.
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn metadata(mut self, metadata: Vec<u8>) -> Self {
+		self.offer.metadata = Some(metadata);
+		self
+	}
+
+	/// Sets the amount for the offer.
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn amount(mut self, amount: Amount) -> Self {
+		self.offer.amount = Some(amount);
+		self
+	}
+
+	/// Sets the features for the offer.
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn features(mut self, features: OfferFeatures) -> Self {
+		self.offer.features = Some(features);
+		self
+	}
+
+	/// Sets the absolute expiry for the offer as seconds since the Unix epoch.
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn absolute_expiry(mut self, absolute_expiry: Duration) -> Self {
+		self.offer.absolute_expiry = Some(absolute_expiry);
+		self
+	}
+
+	/// Sets the issuer for the offer as seconds since the Unix epoch.
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn issuer(mut self, issuer: String) -> Self {
+		self.offer.issuer = Some(issuer);
+		self
+	}
+
+	/// Sets a blinded path for the offer.
+	///
+	/// Successive calls to this method will add another blinded path.
+	pub fn path(mut self, path: BlindedPath) -> Self {
+		self.offer.paths.get_or_insert_with(Vec::new).push(path);
+		self
+	}
+
+	/// Sets a fixed quantity of items for the offer.
+	///
+	/// Successive calls to this method or [`quantity_range`] will override the previous setting.
+	///
+	/// [`quantity_range`]: Self::quantity_range
+	pub fn quantity_fixed(mut self, quantity: NonZeroU64) -> Self {
```

Comment:
Hmm... FWIW, if a user creates an offer with `min_quantity` or `max_quantity` present, then someone scanning it MUST set a quantity in the `invoice_request`. So I'd imagine a wallet would need to implement this, no? Otherwise, it wouldn't be able to create a valid `invoice_request` from a valid `offer`.